python/ray/train/trainer.py

```diff
@@ -330,12 +335,22 @@ def run(
                 checkpoint_strategy=checkpoint_strategy,
                 run_dir=self.latest_run_dir,
             )
+            aggregated_results = collections.defaultdict(list)
+            if aggregate_funcs is None or len(aggregate_funcs) == 0:
+                aggregate_funcs = {}
+            elif isinstance(aggregate_funcs, list):
+                aggregate_funcs = {e.__name__: e for e in aggregate_funcs}
+
             for intermediate_result in iterator:
+                for aggregate_name, func in aggregate_funcs.items():
+                    aggregated_results[aggregate_name].append(func(intermediate_result))
                 for callback in callbacks:
                     callback.process_results(intermediate_result)
 
             assert iterator.is_finished()
-            return iterator.get_final_results()
+            final_results = iterator.get_final_results()
+            final_results.append(aggregated_results)
```

Comment:
Appending `aggreated_results` to `final_results` makes `len(final_results) == num_workers` invalid, which causes failures in test examples.